Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -1159,13 +1159,87 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+
+    auto indexRange = [&] {
+        if (searchDirection == AnnotationSearchDirection::Forward)
+            return [NSIndexSet indexSetWithIndexesInRange:NSMakeRange(startingIndex, [annotations count] - startingIndex)];
+        return [NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, startingIndex + 1)];
+    }();
+
+    auto searchResult = [annotations indexOfObjectAtIndexes:indexRange options:searchDirection == AnnotationSearchDirection::Forward ? 0 : NSEnumerationReverse passingTest:^BOOL(PDFAnnotation* annotation, NSUInteger index, BOOL *stop) {
+        return [annotation isKindOfClass:getPDFAnnotationTextWidgetClass()] && ![annotation isReadOnly] && [annotation shouldDisplay];
+    }];
+
+    return searchResult != NSNotFound ? [annotations objectAtIndex:searchResult] : nullptr;
+}
+
+RetainPtr<PDFAnnotation> UnifiedPDFPlugin::getNextTextAnnotation(AnnotationSearchDirection searchDirection) const
+{
+    ASSERT(m_activeAnnotation);
+    RetainPtr<PDFAnnotation> currentAnnotation = m_activeAnnotation->annotation();
+    RetainPtr<PDFPage> currentPage = [currentAnnotation page];
+    if (!currentPage)
+        return nullptr;
+
+    RetainPtr<NSArray> annotationsForCurrentPage = [currentPage annotations];
+    auto indexOfCurrentAnnotation = [annotationsForCurrentPage indexOfObject:currentAnnotation.get()];
+    ASSERT(indexOfCurrentAnnotation != NSNotFound);
+    if (indexOfCurrentAnnotation == NSNotFound)
+        return nullptr;
+
+    auto isForwardSearchDirection = searchDirection == AnnotationSearchDirection::Forward;
+    if ((isForwardSearchDirection && indexOfCurrentAnnotation + 1 < [annotationsForCurrentPage count]) || (!isForwardSearchDirection && indexOfCurrentAnnotation)) {
+        auto startingIndexForSearch = isForwardSearchDirection ? indexOfCurrentAnnotation + 1 : indexOfCurrentAnnotation - 1;
+        if (RetainPtr<PDFAnnotation> nextTextAnnotationOnCurrentPage = getFirstTextAnnotationStartingAtIndex(annotationsForCurrentPage, startingIndexForSearch, searchDirection))
+            return nextTextAnnotationOnCurrentPage;
+    }
+
+    auto indexForCurrentPage = m_documentLayout.indexForPage(currentPage);
+    if (!indexForCurrentPage)
+        return nullptr;
+
+    RetainPtr<PDFAnnotation> nextAnnotation = nullptr;
+    auto nextPageToSearchIndex = indexForCurrentPage.value();
+    while (!nextAnnotation) {
+        auto getNextPageToSearchIndex = [this, isForwardSearchDirection](unsigned currentPageIndex) -> unsigned {
+            auto pageCount = m_documentLayout.pageCount();
+            if (!isForwardSearchDirection && !currentPageIndex)
+                return pageCount - 1;
+            return isForwardSearchDirection ? ((currentPageIndex + 1) % pageCount) : currentPageIndex - 1;
+        };
+        nextPageToSearchIndex = getNextPageToSearchIndex(nextPageToSearchIndex);
+        RetainPtr<PDFPage> nextPage = m_documentLayout.pageAtIndex(nextPageToSearchIndex);
+        if (!nextPage)
+            continue;
```

Comment:
Shouldn't we bail out here?

Reply:
Maybe.... I changed it but not really sure what the "correct," decision is in this particular scenario